fe/src/main/java/org/apache/doris/transaction/GlobalTransactionMgr.java

```diff
@@ -769,113 +782,48 @@ public boolean isPreviousTransactionsFinished(long endTransactionId, long dbId)
-                        }
-                    }
-                } else {
-                    // check if job is also deleted
-                    // streaming insert stmt not add to fe load job, should use this method to
-                    // recycle the timeout insert stmt load job
-                    if (transactionState.getTransactionStatus() == TransactionStatus.PREPARE
-                            && currentMillis - transactionState.getPrepareTime() > transactionState.getTimeoutMs()) {
-                        if ((transactionState.getSourceType() != LoadJobSourceType.FRONTEND
-                                || !checkTxnHasRelatedJob(transactionState, dbIdToTxnIds))) {
-                            transactionState.setTransactionStatus(TransactionStatus.ABORTED);
-                            transactionState.setFinishTime(System.currentTimeMillis());
-                            transactionState.setReason("transaction is timeout and is cancelled automatically");
-                            unprotectUpsertTransactionState(transactionState);
-                            abortedTxns.add(transactionState);
-                        }
-                    }
+                if (transactionState.isExpired(currentMillis)) {
+                    // remove the txn which labels are expired
+                    expiredTxns.add(transactionState.getTransactionId());
+                } else if (transactionState.isTimeout(currentMillis)) {
+                    // txn is running but timeout, abort it.
+                    timeoutTxns.add(transactionState.getTransactionId());
                 }
             }
-            
-            for (Long transId : transactionsToDelete) {
-                deleteTransaction(transId);
-                LOG.info("transaction [" + transId + "] is expired, remove it from transaction table");
-            }
         } finally {
             writeUnlock();
         }
 
-        for (TransactionState abortedTxn : abortedTxns) {
+        // delete expired txns
+        for (Long txnId : expiredTxns) {
+            deleteTransaction(txnId);
+            LOG.info("transaction [" + txnId + "] is expired, remove it from transaction manager");
+        }
+
+        // abort timeout txns
+        for (Long txnId : timeoutTxns) {
             try {
-                abortedTxn.afterStateTransform(TransactionStatus.ABORTED, true, abortedTxn.getReason());
+                abortTransaction(txnId, "timeout by txn manager");
+                LOG.info("transaction [" + txnId + "] is timeout, abort it by transaction manager");
             } catch (UserException e) {
-                // just print a log, it does not matter.
-                LOG.warn("after abort timeout txn failed. txn id: {}", abortedTxn.getTransactionId(), e);
+                // abort may be failed. it is acceptable. just print a log
+                LOG.warn("abort timeout txn {} failed. msg: {}", txnId, e.getMessage());
```

Comment:
```suggestion
                LOG.warn("abort timeout txn {} failed. msg: {}", txnId, e.getMessage(), e);
```

Reply:
print stack trance is unnecessary here, and will make fe.log ugly